File: lc_go/lcGo/dp/maximumCostSubstring.go
package main

func maximumCostSubstring(s, chars string, vals []int) (ans int) {
	mapping := [26]int{}
	for i := range mapping {
		mapping[i] = i + 1
	}
	for i, c := range chars {
		mapping[c-'a'] = vals[i]
	}
	f := 0
	for _, c := range s {
		f = max(f, 0) + mapping[c-'a']
		ans = max(f, ans)
	}
	return
}

func main() {

}

File: lc_cpp/lcCpp/dp/maximumCostSubstring.cpp
#include <numeric>
//
// Created by 24276 on 2025/8/12.
//
#include <string>
#include <vector>
using namespace std;
class maximum {
public:
    int maximumCostSubstring(string s, string chars, vector<int>& vals) {
        int mapping[26]{};
        std::iota(mapping, mapping + 26, 1);
        for (int i = 0; i < chars.length(); i++) {
            mapping[chars[i] - 'a'] = vals[i];
        }

        // 最大子数组和（允许子数组为空）
        int ans = 0, f = 0;
        for (char c : s) {
            f = max(f, 0) + mapping[c - 'a'];
            ans = max(ans, f);
        }
        return ans;
    }
};
